### ISSUE TITLE
Access token not set correctly after redirect

### ISSUE THREAD

**Issue by 3askaal** [spotify/spotify-web-api-ts-sdk#47]
I have problems making the authentication flow work on the client side. I tried all different approaches but basically the accessToken is always undefined after the initial redirect.

Right now i'm just doing it the same way as in the React example, I just call authenticate on button click instead of pageload. https://github.com/3askaal/collabify/blob/spotify-native-sdk-refactor/client/hooks/useSpotifyApi.ts#L17-L42

If I force authentication a second time it is getting set correctly though. But thats not ideal.

**Comment by sspenst** [comment]
I currently have my code set up like this:

```ts
const [spotifyApi, setSpotifyApi] = useState<SpotifyApi | null>();
```

```ts
const api = SpotifyApi.withUserAuthorization(clientId, redirectUri, scopes);

api.authenticate().then(
  async () => {
    const accessToken = await api.getAccessToken();

    // don't set the spotify api until we have confirmed we are actually logged in
    setSpotifyApi(accessToken ? api : undefined);
  },
  () => setSpotifyApi(null),
);
```

On the first call to `authenticate`, `accessToken` will be null and you will get redirected to the login page. After you get redirected back to your site you should get authenticated correctly.

**Comment by sspenst** [comment]
Also, the reason I am calling `getAccessToken` instead of using the access token from `onfulfilled?: ((value: AccessToken) => void` is because the authenticate function completes before the redirect has finished, so authenticate returns the fake [`emptyAccessToken`](https://github.com/spotify/spotify-web-api-ts-sdk/blob/main/src/auth/AuthorizationCodeWithPKCEStrategy.ts#L65). A fix for this could be to update `redirectOrVerifyToken` to return null when the redirect occurs.

**Comment by 3askaal** [comment]
@sspenst thanks for the reply! I also did try this already though, I'm testing it again right now and this also redirects me to the spotify login page twice in a row. How do you prevent that?

**Comment by sspenst** [comment]
I'm not sure if there is a way to avoid this, since you need to authenticate through Spotify's site and bring that auth code back to your own site. I've only tested the PKCE flow though (`SpotifyApi.withUserAuthorization`), so I can't speak for any of the other auth flows.

Another thing to look into could be the `code` query param that is sent back with your redirect url (eg `http://localhost:3000?code=asdf`). This param is used by `authenticate` and then is removed from the url.

**Comment by 3askaal** [comment]
@sspenst does your application have the same behaviour though? Could be a rendering issue on my side.

And for the `code` I want to ask @thisisjofrank since she is working on this alot, is there a method where I can grant the accessToken from the `code` query param? I looked for it but can't find anything useful.

**Comment by grantshandy** [comment]
> [@3askaal]: ...and only after the second redirect `getAccessToken` returns an actual access token instead of `null`. Really curious how to fix this.

I'm using `withUserAuthorization` and getting this same behavior. It's not until the second time it redirects/authenticates that the token is set, which means the user has to click "accept" on the auth page 2 different times after three redirects, not sure why it's happening either.

**Comment by sspenst** [comment]
@3askaal @grantshandy I made a PR to fix the example React app: #50 

Setting up your app in a similar way should hopefully fix the double auth issue. I tested locally and confirmed it works with Chrome and Safari.

**Comment by 3askaal** [comment]
@sspenst I tried this but it doesnt do the trick for me, the difference here is that everything is in one useEffect, but still get those 2 redirects in a row sadly.

**Comment by sspenst** [comment]
That is very strange... If it didn't work for you using the exact same React code then I'm out of ideas. I would have to see your code to comment further.

**Comment by 3askaal** [comment]
@sspenst are you sure it also works first try when you clear your localStorage?

curious if you can see anything in my code that might cause it too tho: https://github.com/3askaal/collabify/blob/51f9e8238925da2ea72a02e900c4315f428b9a78/client/hooks/useSpotifyApi.ts

**Comment by sspenst** [comment]
@3askaal ok I finally realized how to repro the issue. I'm getting the double auth when running `npm run dev` but not when running `npm run build; npm run preview`. I'll investigate a bit more

**Comment by sspenst** [comment]
Just pushed an update to #50 that fixes the issue in dev mode. It feels not ideal but it works for now...

What I did was add a

```ts
const [mounted, setMounted] = useState(false);
```

and then I wait until the app is "mounted" before attempting to authorize the SpotifyApi:

```ts
  useEffect(() => {
    if (!mounted) {
      setMounted(true);
      return;
    }

    const api = SpotifyApi.withUserAuthorization(
      import.meta.env.VITE_SPOTIFY_CLIENT_ID,
      import.meta.env.VITE_REDIRECT_TARGET,
      Scopes.all,
    );

    api.authenticate().then(async () => {
      const accessToken = await api.getAccessToken();

      if (accessToken) {
        setSpotifyApi(api);
      }
    });
  }, [mounted]);
```

**Comment by grantshandy** [comment]
@3askaal, @sspenst  I do believe this is an *actual bug* in the *source*, not just something in the example. I think I've been able to reproduce it in a simple example without react. I really got into investigating what's going on with this bug 😅 , sorry if it's too long of a read ;)

I think I've found the root of the bug/behavior, here are the steps I took reproduce/find it:

# Test with just `getAccessToken`
```html
<!-- omitted boilerplate -->
<body>
  <p>Logged Out</p>
  <script type="module">
    import {SpotifyApi} from "@spotify/web-api-ts-sdk";

    const sdk = SpotifyApi.withUserAuthorization(
      "<your clientId>",
      window.location.origin,
      [],
    );

    const tokenExistsInStorage = () =>
      (localStorage.getItem('spotify-sdk:AuthorizationCodeWithPKCEStrategy:token') ? true : false);

    console.log('before getAccessToken: ' + tokenExistsInStorage());

    sdk.getAccessToken().then((token) => {
      console.log('after getAccessToken: ' + tokenExistsInStorage());

      if (!token) {
        sdk.authenticate();
      } else {
        document.documentElement.innerHTML = `
          <p>Logged In</p>
          <button onclick="localStorage.clear(); location.reload();">Log Out</button>
        `;
      }
    });
  </script>
</body>
```

## Results
 1. Page loads and redirects you Spotify login page, it redirects you back after the login (expected).
 2. You login and it redirects you back, where you are still logged out?
     - The token exists in `localStorage` on page load, but it is removed after you access `SpotifyApi.getAccessToken()`?
![broken](https://github.com/spotify/spotify-web-api-ts-sdk/assets/45475651/a8f8a8b6-8f7e-40c6-9bca-a4e124068d9f)
 3. Reload the page (or re-authenticate a second time, in @3askaal's case), and everything works correctly:
![fixed](https://github.com/spotify/spotify-web-api-ts-sdk/assets/45475651/3e51e245-ef24-4461-97f2-8845150dd112)
And in my experience this effects all browsers (blink/gecko/webkit):
![image](https://github.com/spotify/spotify-web-api-ts-sdk/assets/45475651/6ff1eff0-42d4-46da-8115-fcb1fd5b83e3)


# Test with an API request
Now add any simple API request(s) after the authentication code in the previous example:
```javascript
sdk
  .currentUser
  .profile()
  .then((profile) => {
    document.documentElement.innerHTML += `<p>User: ${profile.display_name}</p>`;
  });
```

Now we get the behavior @3askaal described where you must authenticate twice before the page works (the page redirects to the auth page twice). I think this is what is happening:
 1. The page redirects to the auth page because no token is found (expected).
 2. On return, `SpotifyApi.getAccessToken` removes the token from memory/`localStorage` (as observed in the first test).
 3. `SpotifyApi.currentUser.profile` detects that there is no token, so automatically runs `SpotifyApi.authenticate` (which redirects).
 4. Once the second auth page redirects back, it resets the "phantom token" bug, and `SpotifyApi.getAccessToken` no longer removes the token. This is equivalent to refreshing in the first test.
 
I think this is why you have to authenticate twice 😅 .
 
**The biggest problem here is that the auth page redirect only resets/fixes the token bug (step 4) on blink browsers. On gecko/webkit browsers, this leads to an infinite redirect/authenticate loop.**
[Screencast](https://github.com/spotify/spotify-web-api-ts-sdk/assets/45475651/f8eed783-3f7e-4d4a-b3a0-50861af2f737)

(this is the bug that I came to this issue for, people can't use my site on firefox/safari)

# Trying API calls without `SpotifyApi.getAccessToken`
If you try the example without the `Spotify.getAccessToken` code...
```javascript
const sdk = SpotifyApi.withUserAuthorization(
  "<your clientId>",
  window.location.origin,
  [],
);

document.documentElement.innerHTML = `
  <p>Logged In</p>
  <button onclick="localStorage.clear(); location.reload();">Log Out</button>
`;

sdk
  .currentUser
  .profile()
  .then((profile) => {
    document.documentElement.innerHTML += `<p>User: ${profile.display_name}</p>`;
  });
```

You get the same behavior, but it reveals an error on the first pageload (before reloading, after which there is no error):
![image](https://github.com/spotify/spotify-web-api-ts-sdk/assets/45475651/22fe98f0-e7ab-4e0d-a599-599a17868345)
```
Error: Failed to refresh token: Bad Request, {"error":"invalid_request","error_description":"refresh_token must be supplied"}
```

I think this is the core of the bug, I'm not sure what to do/where to start in the SDK's source though. The error is coming from [GenericCache.ts](https://github.com/spotify/spotify-web-api-ts-sdk/blob/main/src/caching/GenericCache.ts).

**Comment by davidwhitney** [comment]
@grantshandy hey - I helped out with some of the first implementation of the login/redirect code, don't suppose you could push your repro to GitHub somewhere so I can step through and work out what's going on.

There *is* a feature in the Generic Cache that can remove tokens on access - it shouldn't be in play here other than for the PKCE verification, but where there's smoke there's fire. 10/10 debugging above!

**Comment by sspenst** [comment]
Oh interesting, in that case my addition of [`getAccessToken`](https://github.com/spotify/spotify-web-api-ts-sdk/blob/main/src/SpotifyApi.ts#L147-L149) may be causing this bug + some confusion. My attempt with this function was to expose a way to look at the cache (so you don't have to do something like `localStorage.getItem('spotify-sdk:AuthorizationCodeWithPKCEStrategy:token')`), but I did not realize there are side effects in the [`GenericCache.get`](https://github.com/spotify/spotify-web-api-ts-sdk/blob/main/src/caching/GenericCache.ts#L40) function.

**Comment by sspenst** [comment]
Actually I think I've found the root cause. The example react app is using `React.StrictMode`, which causes useEffects to be loaded twice.
- The first call to `authenticate` removes the `code` from the params [here](https://github.com/spotify/spotify-web-api-ts-sdk/blob/main/src/auth/AuthorizationCodeWithPKCEStrategy.ts#L52-L68)
- The second call is causing the redirect because the `code` has already been removed from the URL

**Comment by grantshandy** [comment]
@davidwhitney I just watched the talk you gave about minimalism in code last week, what a fun coincidence :)

I've set up an example at https://github.com/grantshandy/spotify-auth-debugging [[live](https://grantshandy.github.io/spotify-auth-debugging/)] using my client ID.

**Comment by sspenst** [comment]
@grantshandy thanks for the example! I ran your code using the fix in #53 and it no longer does the double auth, however the strangeness with the before/after getAccessToken still exists. I did some debugging and figured out what's going on:

- On initial page load localStorage is empty and the sdk has no token, so we call `sdk.authenticate()`
- This calls `getOrCreateAccessToken`, which calls `redirectOrVerifyToken` and redirects the user to the auth page
- **Before the redirect has completed, the returned `emptyAccessToken` is set in the cache**
- On redirect back to the page, localStorage has the `emptyAccessToken` and prints `before getAccessToken: true`
- `getAccessToken` returns the `emptyAccessToken`, which also clears it from the cache because it has an almost immediate expiry
- localStorage no longer has any token and prints `after getAccessToken: false`

**TL;DR**: returning `emptyAccessToken` in `redirectOrVerifyToken` is causing bugs because the token value actually gets used

Probably a good fix for this would be to change the return null instead of `emptyAccessToken`, and then make sure that null value is handled correctly throughout the rest of the codebase.

**Comment by 3askaal** [comment]
Thanks alot for the input guys. I don't really have time to dig deeper until next monday but I feel it's pretty clear how it can be fixed now. @sspenst I'm using nextjs on the frontend and im 100% sure I have strict more turned off so that's not the issue for me tho.

**Comment by davidwhitney** [comment]
@sspenst re: "Probably a good fix for this would be to change the return null instead of emptyAccessToken, and then make sure that null value is handled correctly throughout the rest of the codebase."

The idea behind that `emptyAccessToken` is that there is effectively a constant object that can be compared to handle that "null" scenario as I'm pretty retescent to overload what null means - but you're right, half done there, there should be a few guard clauses that check for that `emptyAccessToken` and short circuit this behaviour.

I'll take a crack at it this eve :)

**Comment by davidwhitney** [comment]
Alright, think I've got a fix for this cooked up.

@3askaal I managed to use your repro and have it edited down to

```html

<body>
  <p>Logged Out</p>
  <script type="module">
    import {SpotifyApi} from "@spotify/web-api-ts-sdk"

    const sdk = SpotifyApi.withUserAuthorization(
      "your-app-id-here",
      window.location.origin + window.location.pathname,
      [],
    );
    
    const { authenticated } = await sdk.authenticate();

    if (authenticated) {
      document.documentElement.innerHTML = `
        <p>Logged In</p>
        <button onclick="localStorage.clear(); location.reload();">Log Out</button>
      `;
      
      const profile = await sdk.currentUser.profile();
      document.documentElement.innerHTML += `<p>User: ${profile?.display_name}</p>`;
    }      
  </script>
</body>
```

In a correctly working manner with the new code.

You'll notice

```js
const { authenticated } = await sdk.authenticate();
```
as I've modified the `authenticated` function to do a few checks against the access token to make sure it's both not the emptyAccessToken, and also not expired.

I think we'll be able to cook up a definitive react hook using this and @sspenst's other changes to avoid the react dev server reload stuff. PR incoming 💪🏻

**Comment by thisisjofrank** [comment]
Hey @sspenst does this fix supersede #53 ?

**Comment by 3askaal** [comment]
@davidwhitney for me it works now btw. great work on the fix! I'll leave this open as there was more stuff going on.

**Comment by fostertheweb** [comment]
I am still experiencing the same issue described here with the mixed auth flow using `performUserAuthorization` in SvelteKit. I have a simple login component that calls this method when a button is clicked and the `postbackUrl` receives an empty access token object.
